### PR TITLE
A4A: upgrade callback

### PR DESCRIPTION
**Pull request by keithwrightbos**
Fixes https://github.com/ampproject/amphtml/issues/3976.

### CODE REVIEW

File: extensions/amp-ad/0.1/amp-ad.js [hunk 1]
```diff
@@ -65,21 +44,26 @@ export class AmpAd extends AMP.BaseElement {
     }
     // TODO(tdrl): Check amp-ad registry to see if they have this already.
     if (!a4aRegistry[type] ||
-        !a4aRegistry[type](this.win, this.element)) {
+        !a4aRegistry[type](this.getWin(), this.element)) {
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad.js]: I believe we just use `this.win` now.

Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad.js]: Same here, #4115

File: extensions/amp-ad/0.1/amp-ad.js [hunk 2]
```diff
@@ -65,21 +44,26 @@ export class AmpAd extends AMP.BaseElement {
...
       return new AmpAd3PImpl(this.element);
     }
-    // TODO(dvoytenko): Reimplement a4a via `upgradeCallback`. It will look
-    // as following:
-    /*
-      const extensions = extensionsFor(this.win);
-      return extensions.loadElementClass(extensionTag).then(ctor => {
+    const extensionTagName = networkImplementationTag(type);
+    this.element.setAttribute('data-a4a-upgrade-type', extensionTagName);
+    return extensionsFor(this.getWin()).loadElementClass(extensionTagName)
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad.js]: I've always been confused what file provides the `amp-ad-network-${type}-impl` element class...

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 3]
```diff
@@ -198,14 +198,14 @@ export class AmpA4A extends AMP.BaseElement {
...
-   * Returns true if this element is the child of an amp-ad element.  For use by
+   * Returns true if this element was loaded from an amp-ad element.  For use by
    * network-specific implementations that don't want to allow themselves to be
    * embedded directly into a page.
    * @return {boolean}
    */
   isInAmpAdTag() {
-    return !!this.element.parentElement &&
-        this.element.parentElement.tagName == 'AMP-AD';
+    return this.element.tagName == 'AMP-AD' ||
```
Comment (dvoytenko) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Sounds like this is always true now. Right?

File: extensions/amp-ad/0.1/amp-ad.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-ad/0.1/amp-ad.js]: Correct, `this.win` should be used everywhere now.

File: extensions/amp-ad/0.1/amp-ad.js [hunk 4]
```diff
@@ -65,21 +44,26 @@ export class AmpAd extends AMP.BaseElement {
...
+      .then(ctor => {
+        console.log('upgraded', type, extensionTagName, ctor);
```
Comment (dvoytenko) [inline extensions/amp-ad/0.1/amp-ad.js]: remove

File: extensions/amp-ad/0.1/amp-ad.js (hunk 2 above)
Comment (keithwrightbos) [inline extensions/amp-ad/0.1/amp-ad.js]: It matches an extension (so far amp-ad-network-adsense-impl and amp-ad-network-doubleclick-impl)

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 3 above)
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Not necessarily.  Other ad networks may allow for placing the impl extension (e.g. amp-ad-network-${type}-impl element) directly on the publisher page.  We want to make sure that this is not supported for adsense/doubleclick thus this check.

File: extensions/amp-ad/0.1/amp-ad.js (hunk 1 above)
Comment (keithwrightbos) [inline extensions/amp-ad/0.1/amp-ad.js]: Done

**Comment by keithwrightbos** [comment]
Comments addressed.  Back to @jridgewell & @dvoytenko  Thanks!

File: extensions/amp-ad/0.1/amp-ad.js (hunk 2 above)
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad.js]: Ahh, ok.

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 3 above)
Comment (jridgewell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Shouldn't these always be a `<amp-ad type="X">`? As in, shouldn't directly embedding a `<amp-ad-network-${type}-impl>` be forbidden by the validator?

Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Currently it does not and the requirement to use the loader has some, albeit not less after upgradeCallback support, cost.  I think we should open that discussion to a broader audience to get an idea of what other ad networks may prefer.

**Comment by dvoytenko** [comment]
@keithwrightbos Looks good from my superficial run through. Deferring to @jridgewell and @lannka to do the final approval.

**Comment by jridgewell** [comment]
LGTM.

File: extensions/amp-ad/0.1/amp-ad.js [hunk 5]
```diff
@@ -71,15 +50,19 @@ export class AmpAd extends AMP.BaseElement {
...
     }
-    // TODO(dvoytenko): Reimplement a4a via `upgradeCallback`. It will look
-    // as following:
-    /*
-      const extensions = extensionsFor(this.win);
-      return extensions.loadElementClass(extensionTag).then(ctor => {
+    const extensionTagName = networkImplementationTag(type);
+    this.element.setAttribute('data-a4a-upgrade-type', extensionTagName);
+    return extensionsFor(this.win).loadElementClass(extensionTagName)
+      .then(ctor => {
```
Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad.js]: maybe a good place to inline:
```
.then(ctor => new ctor(this.element))
.catch ...
```

File: extensions/amp-ad/0.1/amp-ad.js [hunk 6]
```diff
@@ -89,20 +72,9 @@ export class AmpAd extends AMP.BaseElement {
...
-    // Note: The loadExtension method will pick the version number.
-    // If we ever reach a point at which there are different extensions with
-    // different version numbers at play simultaneously, we'll have to make sure
-    // that the loader can handle the case.
-    const extensionTag = networkImplementationTag(type);
-    const newChild = this.element.ownerDocument.createElement(extensionTag);
-    extensionsFor(this.win).loadExtension(extensionTag);
-    copyAttributes(this.element, newChild);
-    this.element.appendChild(newChild);
+    // This is only called when no type was set on the element and thus
```
Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad.js]: not sure why don't we fail at the first place?

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 7]
```diff
@@ -198,14 +198,14 @@ export class AmpA4A extends AMP.BaseElement {
   }
 
   /**
-   * Returns true if this element is the child of an amp-ad element.  For use by
+   * Returns true if this element was loaded from an amp-ad element.  For use by
    * network-specific implementations that don't want to allow themselves to be
    * embedded directly into a page.
    * @return {boolean}
    */
   isInAmpAdTag() {
```
Comment (lannka) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Nit pick: not sure if the name is still accurate. Maybe isAmpAdElement()?

File: extensions/amp-ad/0.1/amp-ad.js [hunk 8]
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+/* Copyright 2015 The AMP HTML Authors. All Rights Reserved.
```
Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad.js]: wanna travel back? :-)

File: extensions/amp-ad/0.1/test/test-amp-ad.js [hunk 9]
```diff
@@ -89,33 +90,62 @@ describe('A4A loader', () => {
...
           };
           const element = doc.createElement(tag);
           element.setAttribute('type', 'zort');
-          element.setAttribute('width', '300');
-          element.setAttribute('height', '200');
           doc.body.appendChild(element);
+          const zortInstance = {};
+          const zortConstructor = function() { return zortInstance; };
+          extensionsMock.expects('loadElementClass')
+            .withExactArgs('amp-ad-network-zort-impl')
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad.js]: please keep indentation of 4 spaces for line breaks. 
Applies to the whole file.

File: extensions/amp-ad/0.1/test/test-amp-ad.js [hunk 10]
```diff
@@ -89,33 +90,62 @@ describe('A4A loader', () => {
...
+            expect(
+              doc.head.querySelector('script[src*="amp-ad-network-zort-impl"]'))
+              .to.not.be.null;
+            expect(baseElement).to.equal(zortInstance);
+          });
+        });
+      });
+
+      it('falls back to 3p impl on upgrade with loadElementClass error', () => {
+        return createIframePromise().then(fixture => {
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad.js]: can we try to extract element creation code to `beforeEach()` so we don't repeat them in each test case?

File: extensions/amp-ad/0.1/test/test-amp-ad.js [hunk 11]
```diff
@@ -89,33 +90,62 @@ describe('A4A loader', () => {
...
-              })).to.be.null;
-          expect(expectedChild).to.not.be.null;
-          expect(expectedChild.getAttribute('type')).to.equal('zort');
-          expect(expectedChild.getAttribute('width')).to.equal('300');
-          expect(expectedChild.getAttribute('height')).to.equal('200');
+          const upgradedElementPromise = handler.upgradeCallback();
+          extensionsMock.verify();
+          expect(upgradedElementPromise).not.to.be.null;
+          expect(element.getAttribute(
+            'data-a4a-upgrade-type', 'amp-ad-network-zort-impl'));
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad.js]: is this `expect` check incomplete?

File: extensions/amp-ad/0.1/test/test-amp-ad.js [hunk 12]
```diff
@@ -89,33 +90,62 @@ describe('A4A loader', () => {
...
-              c => {
-                return c.tagName === 'AMP-AD-3P-IMPL';
-              })).to.be.null;
-          expect(expectedChild).to.not.be.null;
-          expect(expectedChild.getAttribute('type')).to.equal('zort');
-          expect(expectedChild.getAttribute('width')).to.equal('300');
-          expect(expectedChild.getAttribute('height')).to.equal('200');
+          const upgradedElementPromise = handler.upgradeCallback();
+          extensionsMock.verify();
+          expect(upgradedElementPromise).not.to.be.null;
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad.js]: this check is probably unnecessary. line 114 implies the check.

File: extensions/amp-ad/0.1/test/test-amp-ad.js [hunk 13]
```diff
@@ -89,33 +90,62 @@ describe('A4A loader', () => {
...
-          expect(expectedChild).to.not.be.null;
-          expect(childElement(element,
-              c => {
-                return c.tagName === 'AMP-AD-3P-IMPL';
-              })).to.be.null;
-          expect(expectedChild).to.not.be.null;
-          expect(expectedChild.getAttribute('type')).to.equal('zort');
-          expect(expectedChild.getAttribute('width')).to.equal('300');
-          expect(expectedChild.getAttribute('height')).to.equal('200');
+          const upgradedElementPromise = handler.upgradeCallback();
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad.js]: is `ampAd` a better name than `handler` here?

File: extensions/amp-ad/0.1/amp-ad.js (hunk 6 above)
Comment (keithwrightbos) [inline extensions/amp-ad/0.1/amp-ad.js]: Done

File: extensions/amp-ad/0.1/amp-ad.js (hunk 5 above)
Comment (keithwrightbos) [inline extensions/amp-ad/0.1/amp-ad.js]: Done

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 7 above)
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Done

File: extensions/amp-ad/0.1/amp-ad.js (hunk 8 above)
Comment (keithwrightbos) [inline extensions/amp-ad/0.1/amp-ad.js]: Ha done... wasn't sure if we should be updating this.

File: extensions/amp-ad/0.1/test/test-amp-ad.js (hunk 9 above)
Comment (keithwrightbos) [inline extensions/amp-ad/0.1/test/test-amp-ad.js]: Done

File: extensions/amp-ad/0.1/test/test-amp-ad.js (hunk 10 above)
Comment (keithwrightbos) [inline extensions/amp-ad/0.1/test/test-amp-ad.js]: Done

File: extensions/amp-ad/0.1/test/test-amp-ad.js (hunk 11 above)
Comment (keithwrightbos) [inline extensions/amp-ad/0.1/test/test-amp-ad.js]: Done

File: extensions/amp-ad/0.1/test/test-amp-ad.js (hunk 13 above)
Comment (keithwrightbos) [inline extensions/amp-ad/0.1/test/test-amp-ad.js]: Done

File: extensions/amp-ad/0.1/test/test-amp-ad.js (hunk 12 above)
Comment (keithwrightbos) [inline extensions/amp-ad/0.1/test/test-amp-ad.js]: Done

**Comment by keithwrightbos** [comment]
lannka@ thanks for the comments.  Back to you!

**Comment by keithwrightbos** [comment]
@lannka ping.  Thanks!

**Comment by lannka** [comment]
Great! LGTM

**Pull request merged**